hbase-coprocessor/src/main/java/org/apache/omid/transaction/TransactionVisibilityFilter.java

```diff
@@ -0,0 +1,244 @@
+            }
+            Long deleteCommit = familyDeletionCache.get(createImmutableBytesWritable(v));
+            if (deleteCommit != null && deleteCommit >= v.getTimestamp()) {
+                if (hbaseTransaction.getVisibilityLevel() == AbstractTransaction.VisibilityLevel.SNAPSHOT_ALL) {
+                    return runUserFilter(v, ReturnCode.INCLUDE_AND_NEXT_COL);
+                } else {
+                    return ReturnCode.NEXT_COL;
+                }
+            }
+            if (CellUtils.isTombstone(v)) {
+                if (hbaseTransaction.getVisibilityLevel() == AbstractTransaction.VisibilityLevel.SNAPSHOT_ALL) {
+                    return runUserFilter(v, ReturnCode.INCLUDE_AND_NEXT_COL);
+                } else {
+                    return ReturnCode.NEXT_COL;
+                }
+            }
+
+            return runUserFilter(v, ReturnCode.INCLUDE_AND_NEXT_COL);
+        }
+
+        return ReturnCode.SKIP;
+    }
+
+
+    private ImmutableBytesWritable createImmutableBytesWritable(Cell v) {
+        return new ImmutableBytesWritable(v.getFamilyArray(),
+                v.getFamilyOffset(),v.getFamilyLength());
+    }
+
+    private ReturnCode runUserFilter(Cell v, ReturnCode snapshotReturn)
+            throws IOException {
+        assert(snapshotReturn == ReturnCode.INCLUDE_AND_NEXT_COL || snapshotReturn == ReturnCode.INCLUDE);
+        if (userFilter == null) {
+            return snapshotReturn;
+        }
+
+        ReturnCode userRes = userFilter.filterKeyValue(v);
+        switch (userRes) {
+            case INCLUDE:
+                return snapshotReturn;
+            case SKIP:
+                return (snapshotReturn == ReturnCode.INCLUDE) ? ReturnCode.SKIP: ReturnCode.NEXT_COL;
+            default:
+                return userRes;
+        }
+
+    }
+
+    // For family delete cells, the sc hasn't arrived yet so get sc from region before going to ct
+    private Optional<Long> getCommitIfInSnapshot(Cell v, boolean getShadowCellBeforeCT) throws IOException {
+        Long cachedCommitTS = commitCache.get(v.getTimestamp());
+        if (cachedCommitTS != null && hbaseTransaction.getStartTimestamp() >= cachedCommitTS) {
+            return Optional.of(cachedCommitTS);
+        }
+        if (snapshotFilter.getTSIfInTransaction(v, hbaseTransaction).isPresent()) {
+            return Optional.of(v.getTimestamp());
+        }
+
+        if (getShadowCellBeforeCT) {
+
+            // Try to get shadow cell from region
+            final Get get = new Get(CellUtil.cloneRow(v));
+            get.setTimeStamp(v.getTimestamp()).setMaxVersions(1);
+            get.addColumn(CellUtil.cloneFamily(v), CellUtils.addShadowCellSuffixPrefix(CellUtils.FAMILY_DELETE_QUALIFIER));
+            Result shadowCell = snapshotFilter.getTableAccessWrapper().get(get);
+
+            if (!shadowCell.isEmpty() &&
+                    Bytes.toLong(CellUtil.cloneValue(shadowCell.rawCells()[0] )) <= hbaseTransaction.getStartTimestamp()){
```

Comment:
Don't repeat Bytes.toLong(CellUtil.cloneValue(shadowCell.rawCells()[0] )) here. You want to minimize heap allocation during filtering.